src/test/java/TestDiscordToMinecraftFormattingTranslator.java

```diff
@@ -0,0 +1,294 @@
+        assertEquals(0, result.peek().start);
+        assertEquals(3, result.peek().end);
+        result.pop();
+        assertTrue(result.empty());
+        assertTrue(translator.getItalicsAst().empty());
+        assertTrue(translator.getBold().empty());
+        assertTrue(translator.getItalicsUnd().empty());
+        assertTrue(translator.getUnderlines().empty());
+        assertTrue(translator.getObfuscated().empty());
+    }
+
+    @Test
+    public void testSimpleBoldItalics() {
+        String message = "***i***";
+        DiscordToMinecraftFormattingTranslator translator = new DiscordToMinecraftFormattingTranslator(message);
+        Stack<DiscordToMinecraftFormattingTranslator.Range> bolds = translator.getBold();
+        Stack<DiscordToMinecraftFormattingTranslator.Range> italics = translator.getItalicsAst();
+
+        assertEquals(0, bolds.peek().start);
+        assertEquals(5, bolds.peek().end);
+        bolds.pop();
+        assertTrue(bolds.empty());
+
+        assertEquals(2, italics.peek().start);
+        assertEquals(4, italics.peek().end);
+        italics.pop();
+        assertTrue(italics.empty());
+    }
+
+    @Test
+    public void testSimpleUnderlineItalics() {
+        String message = "___i___";
+        DiscordToMinecraftFormattingTranslator translator = new DiscordToMinecraftFormattingTranslator(message);
+        Stack<DiscordToMinecraftFormattingTranslator.Range> unds = translator.getUnderlines();
+        Stack<DiscordToMinecraftFormattingTranslator.Range> italics = translator.getItalicsUnd();
+
+        assertEquals(0, unds.peek().start);
+        assertEquals(5, unds.peek().end);
+        unds.pop();
+        assertTrue(unds.empty());
+
+        assertEquals(2, italics.peek().start);
+        assertEquals(4, italics.peek().end);
+        italics.pop();
+        assertTrue(italics.empty());
+    }
+
+    @Test
+    public void testSimpleDisjoints() {
+        String message = "___i___*j*";
+        DiscordToMinecraftFormattingTranslator translator = new DiscordToMinecraftFormattingTranslator(message);
+        Stack<DiscordToMinecraftFormattingTranslator.Range> unds = translator.getUnderlines();
+        Stack<DiscordToMinecraftFormattingTranslator.Range> italicsUnd = translator.getItalicsUnd();
+        Stack<DiscordToMinecraftFormattingTranslator.Range> italicsAst = translator.getItalicsAst();
+
+        assertEquals(0, unds.peek().start);
+        assertEquals(5, unds.peek().end);
+        unds.pop();
+        assertTrue(unds.empty());
+
+        assertEquals(2, italicsUnd.peek().start);
+        assertEquals(4, italicsUnd.peek().end);
+        italicsUnd.pop();
+        assertTrue(italicsUnd.empty());
+
+        assertEquals(7, italicsAst.peek().start);
+        assertEquals(9, italicsAst.peek().end);
+        italicsAst.pop();
+        assertTrue(italicsAst.empty());
+    }
+
+    // @Test
+    // public void testOverlappingItalicsBold0() {
```

Comment:
This is an example of a difference between the DiscordToMinecraftFormattingTranslator's parsing and the official Discord parsing. This test case fails, but I've left it in (commented out) in case a future contributor wants to modify my logic to somehow handle this case. I think I'd have to fundamentally change the parsing approach in order to account for this, but as discussed in the conversation tab I don't think it's that important to have 1:1 perfect Discord formatting representation.